nav2_system_tests/src/costmap_filters/filters_tester.cpp

```diff
@@ -0,0 +1,287 @@
+{
+  RCLCPP_FATAL(get_logger(), "Lifecycle node entered error state");
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+nav2_util::CallbackReturn
+FiltersTester::on_shutdown(const rclcpp_lifecycle::State & state)
+{
+  RCLCPP_INFO(get_logger(), "Shutting down");
+
+  costmap_ros_->on_shutdown(state);
+
+  return nav2_util::CallbackReturn::SUCCESS;
+}
+
+void FiltersTester::startRobotTransform()
+{
+  // Provide the robot pose transform
+  tf_broadcaster_ = std::make_shared<tf2_ros::TransformBroadcaster>(this);
+
+  // Set an initial pose
+  geometry_msgs::msg::Point robot_position;
+  robot_position.x = 0.0;
+  robot_position.y = 0.0;
+  updateRobotPosition(robot_position);
+
+  // Publish the transform periodically
+  transform_timer_ = create_wall_timer(
+    100ms, std::bind(&FiltersTester::publishRobotTransform, this));
+}
+
+void FiltersTester::updateRobotPosition(const geometry_msgs::msg::Point & position)
+{
+  if (!base_transform_) {
+    base_transform_ = std::make_unique<geometry_msgs::msg::TransformStamped>();
+    base_transform_->header.frame_id = "map";
+    base_transform_->child_frame_id = "base_link";
+  }
+
+  base_transform_->header.stamp = now() + rclcpp::Duration(250000000);
+  base_transform_->transform.translation.x = position.x;
+  base_transform_->transform.translation.y = position.y;
+  base_transform_->transform.rotation.w = 1.0;
+
+  publishRobotTransform();
+}
+
+void FiltersTester::publishRobotTransform()
+{
+  if (base_transform_) {
+    tf_broadcaster_->sendTransform(*base_transform_);
+  }
+}
+
+void FiltersTester::spinTester()
+{
+  if (rclcpp::ok()) {
+    // Spin LifycycleNode and Costmap2DROS
+    rclcpp::spin_some(this->get_node_base_interface());
+    rclcpp::spin_some(costmap_ros_->get_node_base_interface());
+  }
+}
+
+bool FiltersTester::testPlan(
+  const geometry_msgs::msg::PoseStamped & start,
+  const geometry_msgs::msg::PoseStamped & end)
+{
+  geometry_msgs::msg::PoseStamped pose = start;
+  nav_msgs::msg::Path path;
+  bool result;
+
+  do {
+    spinTester();
+
+    result = checkPlan(pose, end, path);
+    if (result) {
+      if (path.poses.size() > 2) {
+        // Goal is not reached, continue moving
+        // printPath(path);
+        pose = path.poses[1];
+        if (isInKeepout(pose.pose.position)) {
+          // Fail case: robot enters keepout area
+          return false;
+        }
+        updateRobotPosition(pose.pose.position);
+      } else {
+        // Goal was sucessfully reached
+        return true;
+      }
+      // Sleep for some time to let costmap to be updated
+      std::this_thread::sleep_for(100ms);
+    }
+  } while (result);
+
+  // Fail case: can not produce the path to the goal
+  return false;
+}
+
+bool FiltersTester::isInKeepout(const geometry_msgs::msg::Point & position)
+{
+  const double & x = position.x;
+  const double & y = position.y;
+
+  // Check Bar No.1 on keepout_mask.pgm:
+  if (isInBar(x, y, -4.0, 0.0, -1.7, 1.0)) {
```

Comment:
You didnt actually test the keepout logic if you just hardcoded the range. That's not testing any production code at all then.

Reply:
This is hard-coded values for verification check, not a prerequisite for KeepoutFilter. The input mask for `keepout_filter` is being distributed through a real PGM-file. Anyway, I would like agree, that hard-coded values are not suitable if we want to have general test that might be used by another developers in their code. I will change it in the update for mask or costmap vs. pose verification.